src/main/java/net/pms/dlna/DLNAMediaDatabase.java

```diff
@@ -574,39 +569,49 @@ public synchronized boolean isOpenSubtitlesMetadataExists(String name, long modi
-							audio.setSampleFrequency(elements.getString("SAMPLEFREQ"));
-							audio.setCodecA(elements.getString("CODECA"));
-							audio.setBitsperSample(elements.getInt("BITSPERSAMPLE"));
-							audio.setAlbum(elements.getString("ALBUM"));
-							audio.setArtist(elements.getString("ARTIST"));
-							audio.setSongname(elements.getString("SONGNAME"));
-							audio.setGenre(elements.getString("GENRE"));
-							audio.setYear(elements.getInt("YEAR"));
-							audio.setTrack(elements.getInt("TRACK"));
-							audio.getAudioProperties().setAudioDelay(elements.getInt("DELAY"));
-							audio.setMuxingModeAudio(elements.getString("MUXINGMODE"));
-							audio.setBitRate(elements.getInt("BITRATE"));
-							media.getAudioTracksList().add(audio);
-						}
+					ResultSet elements;
+					audios.setLong(1, id);
+					elements = audios.executeQuery();
+					while (elements.next()) {
+						DLNAMediaAudio audio = new DLNAMediaAudio();
+						audio.setId(elements.getInt("ID"));
+						audio.setLang(elements.getString("LANG"));
+						audio.setAudioTrackTitleFromMetadata(elements.getString("TITLE"));
+						audio.getAudioProperties().setNumberOfChannels(elements.getInt("NRAUDIOCHANNELS"));
+						audio.setSampleFrequency(elements.getString("SAMPLEFREQ"));
+						audio.setCodecA(elements.getString("CODECA"));
+						audio.setBitsperSample(elements.getInt("BITSPERSAMPLE"));
+						audio.setAlbum(elements.getString("ALBUM"));
+						audio.setArtist(elements.getString("ARTIST"));
+						audio.setSongname(elements.getString("SONGNAME"));
+						audio.setGenre(elements.getString("GENRE"));
+						audio.setYear(elements.getInt("YEAR"));
+						audio.setTrack(elements.getInt("TRACK"));
+						audio.getAudioProperties().setAudioDelay(elements.getInt("DELAY"));
+						audio.setMuxingModeAudio(elements.getString("MUXINGMODE"));
+						audio.setBitRate(elements.getInt("BITRATE"));
+						media.getAudioTracksList().add(audio);
 					}
-					subs.setInt(1, id);
-					try (ResultSet elements = subs.executeQuery()) {
-						while (elements.next()) {
-							DLNAMediaSubtitle sub = new DLNAMediaSubtitle();
-							sub.setId(elements.getInt("ID"));
-							sub.setLang(elements.getString("LANG"));
-							sub.setSubtitlesTrackTitleFromMetadata(elements.getString("TITLE"));
-							sub.setType(SubtitleType.valueOfStableIndex(elements.getInt("TYPE")));
-							media.getSubtitleTracksList().add(sub);
+
+					elements.close();
+					subs.setLong(1, id);
+					elements = subs.executeQuery();
+					while (elements.next()) {
+						String fileName = elements.getString("EXTERNALFILE");
+						File externalFile = isNotBlank(fileName) ? new File(fileName) : null;
+						if (externalFile != null && !externalFile.exists()) {
+							LOGGER.trace("Deleting cached external subtitles from database because the file \"{}\" doesn't exist", externalFile.getPath());
+							deleteRowsInTable("SUBTRACKS", "EXTERNALFILE", externalFile.getPath(), false);
```

Comment:
@valib it occurred to me that this might be better suited to go into our database cleanup step instead of happening here, since then it is asynchronous to the request to `getData` and therefore will be more performant. Do you agree?

Reply:
@SubJunk the database cleanup is called when the scannig of shared folders finished but what about the case when user is browsing folders and subs were deleted in the mean time and scanning is not finished?

Reply:
@valib that makes sense, thanks for explaining